### DISCUSSION TITLE
Automatic Formatting

### DISCUSSION THREAD

**Discussion by rteyssier** [ramses-organisation/ramses#51] [Ideas]
Hi everyone,
There is a pull request to allow for automatic Fortran formatting when there is a new PR.
This will improve readability and robustness. 
The downside is that formatting is editor-dependent and you may see changes in the formatting that you may not like.
Before we decide to merge this pull request, we want to make sure automatic formatting is generally considered as a good idea.

**Comment by nbrucy** [comment]
That is PR #44. This will not force people to comply to the formatting and change their coding habits. 
Instead, it will reformat new code when it's submitted via a pull request.
This means that any personal but not standard indenting (like adding whitespaces to align variable names) will be lost.
However this progressively improve readability by a lot.

I think it's globally a good asset and we should have it. However, to reduce the annoyance to contributors to the minimum, I think it would be better that it is not applied when a PR is submitted, but just before it is merged.
That will ensure that people don't get conflict when they are updating their PR. Do you think it's possible @cphyc?

**Reply by cphyc** [reply]
Yes, this is possible to implement, with a draft in #44. 

To make sure this is crystal clear, what I implement in the aforementioned PR ensures that:
- there is nothing to do by the contributor,
- only the files they touch will be formatted,
- this can be triggered by adding a label to the PR, or by commenting some magic keyword by one of the reviewer,
- if the formatting isn't "correct", merging is blocked (in the form of a failing test).

All but point 3 are implemented in #44.

**Reply by nbrucy** [reply]
I think point 3 is important. We may not want to fix the formatting of the entire file if the modification is only a small bugfix, which will be hard to catch in the diff. Also auto-applying commit into the PR can generate conflicts when updating the PR, so to avoid we may want to apply the formatting only at the very end, just before merging.

Also I don't understand point 4: why should the merging blocked if precommit autofixes the PR?

**Reply by cphyc** [reply]
Because we _do not_ want precommit to autofix the PR (it can only do so automatically whenever the PR is issued). So it needs to be done semi-manually, and the blocking is a way of enforcing it.

Formatting cannot be done just when we merge, because it will fail if, e.g., the file isn't proper Fortran (in which case test should also fail, unless we're touching a file that isn't tested). And we also want to keep an eye on formatting in case something goes wrong.

**Reply by nbrucy** [reply]
Ok but currently the precommit hook for trailing whitespace is autofixing any commit added to open PR (which can be a bit annoying).

**Reply by cphyc** [reply]
Yes - we can (should?) deactivate it.

**Reply by nbrucy** [reply]
Maybe. I think manually triggered autofixes are better in our case.

**Reply by cphyc** [reply]
Ah I actually can't do it, only Romain can.

**Reply by tinecolman** [reply]
I can testify it is annoying as it is now. I wanted to update my PR but  the thing complained very dramatically that my branch diverged because of some dumb trailing spaces.

**Comment by cphyc** [comment]
Note that most of the editor-specific configuration is the number of spaces per tab, which for most editors is either 2 or 4, except for emacs where it is (by default) 2 or 3. Note that this can be changed.

Most of the changes come from transforming for example:
```fortran
subroutine foo(arg1,arg2,arg3)
   integer::i,j,k,l,m,n

   if(i.eq.j)then
                  m=n
   endif
end
```

will be formatted to
```fortran
subroutine foo(arg1, arg2, arg3)
   integer :: i, j, k, l, m, n
   
   if (i == j) then
       m = n
   end if
end subroutine foo
```
which is arguably more readable.

**Reply by nbrucy** [reply]
By curiosity (because Joe asked for it), is there an option to get 

```fortran
subroutine foo(arg1, arg2, arg3)
   integer :: i, j, k, l, m, n
   
   if (i == j) then
       m = n
   end if ! (i == j)
end subroutine foo
```
(ie to recall the if condition via a comment everytime their is an end if)

**Reply by cphyc** [reply]
My fortan formatter is very similar in spirit to Black for Python. So although it is technically possible, it isn't implemented. One reason against it is, how do you format the following?

```fortran
if ((a == .true.) .and. &
    (b * b - 4 * a * c < 34 .and. &
    (this_variable == that_variable) .or. &
    (my_horse == amazing) &
) then
   ! do something on the diagonal if a is true and your horse is amazing
end if ! Panic: what do I write here?
```

**Reply by jlewis47** [reply]
> My fortan formatter is very similar in spirit to Black for Python. So although it is technically possible, it isn't implemented. One reason against it is, how do you format the following?
> 
> ```fortran
> if ((a == .true.) .and. &
>     (b * b - 4 * a * c < 34 .and. &
>     (this_variable == that_variable) .or. &
>     (my_horse == amazing) &
> ) then
>    ! do something on the diagonal if a is true and your horse is amazing
> end if ! Panic: what do I write here?
> ```

Yeah this could be very messy! Better for people to just add a quick comment if we think it's a good practice
